File: packages/gavio-js/src/providers/gemini.ts
/** geminiAdapter — Google Generative Language API (generateContent). */

import { ConfigurationError } from '../errors.js'
import type { PricingProvider } from '../pricing.js'
import type { GavioRequest } from '../request.js'
import type { GavioResponse } from '../response.js'
import { TokenUsage, type Message } from '../types.js'
import { BaseProviderAdapter } from './base.js'
import { postJson } from './http.js'

const DEFAULT_BASE_URL = 'https://generativelanguage.googleapis.com/v1beta'

export interface GeminiAdapterOptions {
  apiKey?: string
  baseUrl?: string
  timeoutMs?: number
  pricing?: PricingProvider
}

interface GeminiContent {
  role: string
  parts: { text: string }[]
}

/** Map Gavio messages to Gemini contents + a system instruction. */
export function geminiToContents(messages: Message[]): {
  system: string | null
  contents: GeminiContent[]
} {
  let system: string | null = null
  const contents: GeminiContent[] = []
  for (const m of messages) {
    const text = m.content
    if (m.role === 'system') {
      system = system ? `${system}\n${text}` : text
      continue
    }
    contents.push({ role: m.role === 'assistant' ? 'model' : 'user', parts: [{ text }] })
  }
  return { system, contents }
}

class GeminiAdapter extends BaseProviderAdapter {
  private readonly apiKey: string | undefined
  private readonly baseUrl: string
  private readonly timeoutSeconds: number

  constructor(options: GeminiAdapterOptions = {}) {
    super(options.pricing)
    this.apiKey =
      options.apiKey ?? process.env['GEMINI_API_KEY'] ?? process.env['GOOGLE_API_KEY']
    this.baseUrl = (options.baseUrl ?? DEFAULT_BASE_URL).replace(/\/+$/, '')
    this.timeoutSeconds = (options.timeoutMs ?? 30_000) / 1000
  }

  get providerName(): string {
    return 'gemini'
  }

  private payload(request: GavioRequest): Record<string, unknown> {
    const { system, contents } = geminiToContents(request.messages)
    const payload: Record<string, unknown> = {
      contents,
      generationConfig: {
        temperature: request.temperature,
        maxOutputTokens: request.maxTokens,
      },
    }
    if (system) payload['systemInstruction'] = { parts: [{ text: system }] }
    return payload
  }

  async complete(request: GavioRequest): Promise<GavioResponse> {
    if (!this.apiKey) throw new ConfigurationError('GEMINI_API_KEY not set')
    const started = performance.now()
    const url = `${this.baseUrl}/models/${request.model}:generateContent?key=${this.apiKey}`
    const data = await postJson(url, this.payload(request), {}, this.timeoutSeconds)
    const candidates = (data['candidates'] as Array<Record<string, unknown>>) ?? [{}]
    const contentObj = (candidates[0]?.['content'] as Record<string, unknown>) ?? {}
    const parts = (contentObj['parts'] as Array<{ text?: string }>) ?? []
    const content = parts.map((p) => p.text ?? '').join('')
    const um = (data['usageMetadata'] as Record<string, number>) ?? {}
    const usage = new TokenUsage(um['promptTokenCount'] ?? 0, um['candidatesTokenCount'] ?? 0)
    return this.buildResponse(request, content, usage, request.model, started)
  }

  async healthCheck(): Promise<boolean> {
    return !!this.apiKey
  }
}

/** Factory: build a Gemini provider adapter. */
export function geminiAdapter(options: GeminiAdapterOptions = {}): GeminiAdapter {
  return new GeminiAdapter(options)
}

File: packages/gavio-py/gavio/providers/gemini.py
"""GeminiAdapter — Google Generative Language API (generateContent)."""

from __future__ import annotations

import os
import time

from ..exceptions import ConfigurationError, ProviderUnavailableError
from ..pricing import PricingProvider
from ..request import GavioRequest
from ..response import GavioResponse
from ..types import Message, TokenUsage
from ._http import post_json
from .base import ProviderAdapter

_DEFAULT_BASE_URL = "https://generativelanguage.googleapis.com/v1beta"


class GeminiAdapter(ProviderAdapter):
    """Talks to the Gemini generateContent endpoint over HTTP.

    Gemini uses roles ``user``/``model`` (not ``assistant``) and a separate
    ``systemInstruction`` field, so messages are mapped accordingly.
    """

    def __init__(
        self,
        api_key: str | None = None,
        base_url: str | None = None,
        timeout_seconds: float = 30.0,
        pricing: PricingProvider | None = None,
    ) -> None:
        super().__init__(pricing)
        self._api_key = (
            api_key
            or os.environ.get("GEMINI_API_KEY")
            or os.environ.get("GOOGLE_API_KEY")
        )
        self.base_url = (base_url or _DEFAULT_BASE_URL).rstrip("/")
        self.timeout_seconds = timeout_seconds

    @property
    def provider_name(self) -> str:
        return "gemini"

    @staticmethod
    def _to_contents(
        messages: list[Message],
    ) -> tuple[str | None, list[dict]]:
        system: str | None = None
        contents: list[dict] = []
        for m in messages:
            role = m.get("role")
            text = m.get("content", "")
            if role == "system":
                system = f"{system}\n{text}" if system else text
                continue
            g_role = "model" if role == "assistant" else "user"
            contents.append({"role": g_role, "parts": [{"text": text}]})
        return system, contents

    def _payload(self, request: GavioRequest) -> dict:
        system, contents = self._to_contents(request.messages)
        payload: dict = {
            "contents": contents,
            "generationConfig": {
                "temperature": request.temperature,
                "maxOutputTokens": request.max_tokens,
            },
        }
        if system:
            payload["systemInstruction"] = {"parts": [{"text": system}]}
        return payload

    async def complete(self, request: GavioRequest) -> GavioResponse:
        if not self._api_key:
            raise ConfigurationError("GEMINI_API_KEY not set")
        started = time.monotonic()
        url = (
            f"{self.base_url}/models/{request.model}:generateContent"
            f"?key={self._api_key}"
        )
        data = await post_json(
            url, self._payload(request), {}, timeout=self.timeout_seconds
        )
        candidates = data.get("candidates") or [{}]
        parts = candidates[0].get("content", {}).get("parts", [])
        content = "".join(p.get("text", "") for p in parts)
        um = data.get("usageMetadata", {})
        usage = TokenUsage(
            prompt_tokens=um.get("promptTokenCount", 0),
            completion_tokens=um.get("candidatesTokenCount", 0),
        )
        return self._build_response(request, content, usage, request.model, started)

    async def health_check(self) -> bool:
        try:
            return bool(self._api_key)
        except ProviderUnavailableError:
            return False
